plugins/org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/ui/quickfix/QuickfixProviderFragment2.xtend

```diff
@@ -0,0 +1,177 @@
+import org.eclipse.xtext.Grammar
+import org.eclipse.xtext.xtext.generator.AbstractGeneratorFragment2
+import org.eclipse.xtext.xtext.generator.CodeConfig
+import org.eclipse.xtext.xtext.generator.XtextGeneratorNaming
+import org.eclipse.xtext.xtext.generator.model.FileAccessFactory
+import org.eclipse.xtext.xtext.generator.model.GuiceModuleAccess
+import org.eclipse.xtext.xtext.generator.model.TypeReference
+
+import static extension org.eclipse.xtext.xtext.generator.util.GrammarUtil2.*
+
+/**
+ * Contributes the Quickfix provider stub, either in Xtend or Java language.
+ * 
+ * @author Christian Schneider - Initial contribution and API
+ */
+class QuickfixProviderFragment2 extends AbstractGeneratorFragment2 {
+
+	@Inject
+	extension XtextGeneratorNaming
+
+	@Inject
+	extension CodeConfig
+
+	@Inject
+	FileAccessFactory fileAccessFactory
+
+	@Accessors
+	private boolean generateStub = true;
+
+	@Accessors	
+	private boolean inheritImplementation;
+
+	def protected TypeReference getQuickfixProviderClass(Grammar g) {
+		return new TypeReference(
+			grammar.eclipsePluginBasePackage + ".quickfix." + grammar.simpleName + "QuickfixProvider"
+		)
+	}
+
+	def protected TypeReference getQuickfixProviderSuperClass(Grammar g) {
+		val superGrammar = g.getNonTerminalsSuperGrammar;
+		if (inheritImplementation && superGrammar != null) 
+			superGrammar.quickfixProviderClass
+		else
+			defaultQuickfixProviderSuperClass
+	}
+
+	/**
+	 * Extra getter facilitates customization by overriding. 
+	 */
+	protected def getDefaultQuickfixProviderSuperClass() {
+		return new TypeReference("org.eclipse.xtext.ui.editor.quickfix.DefaultQuickfixProvider")
+	} 
+
+	override generate() {
+		val instanceClass = 
+			if (generateStub) grammar.quickfixProviderClass else grammar.quickfixProviderSuperClass;
+
+		new GuiceModuleAccess.BindingFactory()
+				.addTypeToType(
+					new TypeReference("org.eclipse.xtext.ui.editor.quickfix.IssueResolutionProvider"),
+					instanceClass
+				).contributeTo(language.eclipsePluginGenModule);
+
+		if (!generateStub) {
+			return;
+		}
+
+		if (projectConfig.eclipsePluginSrc !== null) {
+			if (preferXtendStubs) {
+				generateXtendQuickfixProvider
+			} else {
+				generateJavaQuickfixProvider
+			}
+
+			projectConfig.eclipsePluginManifest.exportedPackages += grammar.quickfixProviderClass.packageName
+
+			addRegistrationToPluginXml
+		}
+	}
+
+	protected def generateXtendQuickfixProvider() {
+		fileAccessFactory.createXtendFile(grammar.quickfixProviderClass, '''
+			//import org.eclipse.xtext.ui.editor.quickfix.Fix
+			//import org.eclipse.xtext.ui.editor.quickfix.IssueResolutionAcceptor
+			//import org.eclipse.xtext.validation.Issue
+			
+			/**
+			 * Custom quickfixes.
+			 *
+			 * See https://www.eclipse.org/Xtext/documentation/304_ide_concepts.html#quick-fixes
+			 */
+			class �grammar.quickfixProviderClass.simpleName� extends �grammar.quickfixProviderSuperClass� {
+			
+			//	@Fix(MyDslValidator.INVALID_NAME)
```

Comment:
Should we use the real validator name in this code snippet or is it not important?